backend/search/models.py

```diff
@@ -59,18 +59,41 @@ def _get_cache_path(self) -> pathlib.Path:
         settings.CACHING_DIR.mkdir(exist_ok=True, parents=True)
         return settings.CACHING_DIR / str(self.id)
 
-    def check_results(self) -> bool:
+    def check_cache(self) -> bool:
+        """Check if the cache file is readable and if its contents corresponds
+        to what the database knows about it. If not, the search will have to 
+        be repeated. If search is not yet completed, the cache size is not
+        tested."""
+        cache_path = self._get_cache_path()
         try:
-            self.get_results()
+            with cache_path.open():
+                pass
+        except OSError:
+            return False
+        # Cache file is readable, now check if size is correct
+        if self.search_completed is None:
             return True
-        except Exception:
-            logger.exception('Failed reading results of ComponentSearchQuery: %d', self.pk)
-
-        return False
+        else:
+            actual_size = cache_path.stat().st_size
+            size_equal = actual_size == self.cache_size
+            if not size_equal:
+                logger.warning(
+                    f"Cache size for {self} is {actual_size} while it is supposed to be {self.cache_size}"
+                )
+            return size_equal
 
     def get_results(self) -> ResultSet:
         """Return results as a dict"""
-        results = self._get_cache_path().read_text()
+        cache_path = self._get_cache_path()
+        try:
+            results = cache_path.read_text()
+        except OSError:
+            # Cache file is not present or not readable. This may happen when
+            # results are fetched but searching has not yet started, or if
+            # the cache file was deleted. Go on and return an empty set -
+            # this is not problematic because an inconsistency between model
+            # and cache will be detected by the code that performs the search.
+            results = ''
```

Comment:
might be nicer to `return []`

Reply:
Writing this here as well for reference: I think one of the problems we previously had is that in some cases running a query would give 0 results, and it wasn't clear to the user that an error had occurred (as opposed to the query really having no results)
Please verify that this change isn't reintroducing that behaviour.

Reply:
That behaviour was because we were not regenerating results at the time in case of errors. We would simply record that an error occurred and keep the cache empty. But the perform_search method now properly checks if there were errors for every CSR and redoes the search if so.